### PR TITLE
Complete rcutils API coverage

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says, filling the gaps with mocks. Depends on https://github.com/ros2/Mimick/pull/7.

### CODE REVIEW

**Comment by hidmic** [comment]
This isn't PR material just yet, but I wanted to open it early enough for folks to take a look at how tests are looking like, and, ideally, compare it side-by-side with @brawner's fault injection tooling. There are cases that definitely suit mocking conceptually (e.g. dealing with bad system clocks), others not so much (e.g. failing a finalization).

**Comment by hidmic** [comment]
Total coverage is over 95%, but there're places where it is still below that mark. Those places were left out on purpose, and can be split into three (3) groups:

- Those that can still be tested without mocking (e.g. using a bad allocator).
- Those that I think are good candidates for fault injection (e.g. typical cascading errors during finalization).
- Those that I think could use a revisit e.g. the use of a `rcutils_string_map_t` in `logging` to track logger severity levels adds failure points and allocations/deallocation cycles on every log for no obvious reason.

File: test/mocking_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+#if !defined(_WIN32)
+#if defined(__MACH__)
+
+class fake_clock : unique_in_scope<fake_clock>
+{
+public:
+  fake_clock()
+  : unique_in_scope(this)
+  {
+    mock_ = mmk_mock(RCUTILS_STRINGIFY(clock_get_time) "@lib:rcutils", clock_get_time_mock_t);
```
Comment (brawner) [inline test/mocking_utils.hpp:72]: What's the purpose of RCUTILS_STRINGIFY here instead of just writing out the full string?

File: test/test_char_array.cpp [hunk 2]
```diff
@@ -122,11 +124,21 @@ TEST_F(ArrayCharTest, vsprintf_fail) {
   ASSERT_EQ(RCUTILS_RET_OK, ret);
 
   char_array.allocator = failing_allocator;
-  ret = example_logger(&char_array, "Long string for the case %d", 2);
+  ret = example_logger(&char_array, "Long string for given %s", "array");
```
Comment (brawner) [inline test/test_char_array.cpp:127]: If this change is important, I would add a comment that describes why a %s is needed instead of a %d

File: test/test_format_string.cpp [hunk 3]
```diff
@@ -17,9 +17,12 @@
 #include <string>
 
 #include "./allocator_testing_utils.h"
+
 #include "rcutils/allocator.h"
 #include "rcutils/format_string.h"
 
+#include "./mocking_utils.hpp"
```
Comment (brawner) [inline test/test_format_string.cpp:24]: Would it make sense to group the two relative imports?

File: CMakeLists.txt [hunk 4]
```diff
@@ -181,7 +181,7 @@ if(BUILD_TESTING)
     osrf_testing_tools_cpp::memory_tools LIBRARY_PRELOAD_ENVIRONMENT_IS_AVAILABLE)
 
   ament_add_gtest(test_logging test/test_logging.cpp)
-  target_link_libraries(test_logging ${PROJECT_NAME} osrf_testing_tools_cpp::memory_tools)
+  target_link_libraries(test_logging ${PROJECT_NAME} mimick osrf_testing_tools_cpp::memory_tools)
```
Comment (brawner) [inline CMakeLists.txt:184]: How many libraries need to be added before these calls should be separated into multiple lines?

File: test/mocking_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+  {
+    return ts_;
+  }
+
+private:
+  mmk_mock_define(clock_get_time_mock_t, kern_return_t, clock_serv_t, mach_timespec_t *);
+  static kern_return_t do_clock_get_time(clock_serv_t, mach_timespec_t * ts)
+  {
+    *ts = instance->ts_;
+    return 0;
```
[the remark this answers is not in the document]
Comment (brawner) [reply test/mocking_utils.hpp:93]: To clarify, templating the class might negate the need for separating these classes if the overall code can be significantly reduced in size.

File: src/char_array.c [hunk 6]
```diff
@@ -172,14 +172,14 @@ rcutils_char_array_vsprintf(rcutils_char_array_t * char_array, const char * form
   if (new_size > char_array->buffer_capacity) {
     rcutils_ret_t ret = rcutils_char_array_expand_as_needed(char_array, new_size);
     if (ret != RCUTILS_RET_OK) {
-      RCUTILS_SET_ERROR_MSG("char array failed to expand");
```
[the remark this answers is not in the document]
Comment (hidmic) [reply src/char_array.c:175]: These error messages overwrite the ones already set by `rcutils_char_array_expand_as_needed()`. It is not strictly necessary to fix such error handling details in this PR though.

File: src/char_array.c [hunk 7]
```diff
@@ -172,14 +172,14 @@ rcutils_char_array_vsprintf(rcutils_char_array_t * char_array, const char * form
...
       return ret;
     }
 
     if (_rcutils_char_array_vsprintf(char_array, format, args) != size) {
       if (rcutils_char_array_fini(char_array) == RCUTILS_RET_OK) {
         RCUTILS_SET_ERROR_MSG("vsprintf on resized char array failed");
       } else {
+        rcutils_reset_error();
```
[the remark this answers is not in the document]
Comment (hidmic) [reply src/char_array.c:182]: Same as above. It's not necessary to fix these issues here, but we don't do error handling correctly (i.e. bunch of overwrites here and there).

File: test/mocking_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,415 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:1]: > There is a lot happening in this file. I think it should be broken up into separate files and placed in its own directory, unless there is better consolidation that can happen through templated classes. Feel free to create an all encompassing header that just imports the other ones. Also, it might be helpful for consolidating examples to create a file with unit tests that only test these specific mocking classes.

Makes sense, logic is a bit too sophisticated.

> I think these classes will also need some doxygen style documentation and comments since they are meant to be used by others than just ourselves. It took me awhile reading this file to understand all that was happening.

Agreed as well.

> Also this is a cpp file. What impact might that have if it gets moved out of a test directory? Will it need to be placed in rcpputils instead? Or a new package?

We could move them to `rcpputils`, yes. I didn't initially because these classes are somewhat tailored to `rcutils`' testing needs, and an attempt to concentrate all platform-specific details and boilerplate code in a single place. But yeah, perhaps we find use cases for them elsewhere. 

While we are at it, maybe we can move other testing tools as well (e.g. the bunch of `allocator_testing_utils.h` duplicates we have laying around).

File: test/mocking_utils.hpp [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+  T * prev_instance_ {nullptr};
+};
+
+template<typename T>
+T * unique_in_scope<T>::instance {nullptr};
+
+#if !defined(_WIN32)
+#if defined(__MACH__)
+
+class fake_clock : unique_in_scope<fake_clock>
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:66]: Being an internal testing tool, I didn't see the need to standardize their names. But moving them out into `rcpputils` we have to.

File: test/mocking_utils.hpp [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+    return 0;
+  }
+
+  mach_timespec_t ts_ {};
+  clock_get_time_mock_t mock_;
+};
+
+#else  // defined(__MACH__)
+
+class fake_clock : unique_in_scope<fake_clock>
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:102]: > Can you consolidate these differences to a much smaller #if defined block to reduce duplicated code? 

In this particular case, yes we can. A few `using` directives can go a long way here.

In general, all "equivalent" classes can be consolidated into one, but in some cases I think interleaving all platform-specific details works against readability and thus I tend to split.

File: test/test_char_array.cpp (hunk 2 above)
Comment (hidmic) [reply test/test_char_array.cpp:127]: Not important, a leftover.

File: test/test_format_string.cpp (hunk 3 above)
Comment (hidmic) [reply test/test_format_string.cpp:24]: It would.

File: CMakeLists.txt (hunk 4 above)
Comment (hidmic) [reply CMakeLists.txt:184]: Hmm, I don't think we have a policy -- `lint_cmake` is not very pedantic. But I don't mind wrapping.

File: test/mocking_utils.hpp (hunk 1 above)
Comment (hidmic) [reply test/mocking_utils.hpp:72]: This is a fun detail. System libraries have a good deal of versioning mechanisms, and sometimes the binary API standards (symbol names) do not match the source API standards (identifier names). I've seen two techniques so far: that of using macros for the forward facing API, and that of using function in-lining. This is a counter-measure for the first, I have yet to find one for the later (reason why `glibc`'s `__xstat` is mentioned explicitly below).

File: test/mocking_utils.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:93]: I didn't want to start with full-blown C++ generic wrappers, but yeah, that can be done. I actually have some of that code laying around. Still, I'm not sure I'd mock system library API directly in each test. There're too many platform-specific details and nuances that I'd rather keep in one place, plus some APIs have to show a consistent view (e.g. when faking a filesystem) which IMHO warrants a single unified mocking tool.

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:93]: In short, I can add some generic functionality to `rcpputils` to reduce duplication, but I'd keep system library mocking (and all its quirks) in one place (perhaps alongside).

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:93]: Hmm thinking about this again, it's not going to be pretty. I can't use variadic templates: `mimick` does code generation using macros and so I need argument types (i.e. template parameters) at the preprocessing stage.

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:93]: Scratch that, it's not possible with Mimick as-is. The call to the function to be mocked must be hard-coded to inject the replacement in the first place.

File: test/mocking_utils.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Blast545) [reply test/mocking_utils.hpp:1]: If it gets moved to `rcpputils` wouldn't it add an additional `mimick_vendor` dependency to `rcpputils`? If that's the case, then it might not be a good approach because a lot of packages depend on `rcpputils`

> While we are at it, maybe we can move other testing tools as well
I think that would be a good idea. Maybe having a repo for testing utils? bomb/failure allocators, mock structures and maybe any other needed for performance testing.

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:1]: It will. But it already does for `rcutils` which AFAIK is even more widely used than `rcpputils`.

[the remark this answers is not in the document]
Comment (Blast545) [reply test/mocking_utils.hpp:1]: In this case, it depends on `mimick_vendor` for testing. I was talking about including `mimick_vendor` in `rcpputils` for `rcutils` tests (or others) will imply a direct runtime dependency, am I right?

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:1]: If by optional you mean they won't be pulled in if the library is installed or distributed in binary, yes, that's correct. 

Testing utilities don't have a place yet, we duplicate (which I'd rather stop doing). Whether `rcutils` and/or  `rcpputils` are the right places for them, I don't have a strong preference. Any package name in mind?

File: test/mocking_utils.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:93]: Alright, it wasn't super trivial but I think I managed to do wrap `Mimick` in a more C++ like API. PTAL at https://github.com/ros2/rcpputils/pull/92. Once we agree there, I'll refactor the mocks in this patch on top of it.

File: test/mocking_utils.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Blast545) [reply test/mocking_utils.hpp:1]: > If by optional you mean they won't be pulled in if the library is installed or distributed in binary, yes, that's correct.

Yes, that would be a concern because if `mimick_vendor` becomes a runtime dependency of `rcpputils/rcutils` we'll have to target QL1 for `mimick`.

About the name, I was thinking on something like `testing_utils`, but I don't have strong preferences on it.

[the remark this answers is not in the document]
Comment (brawner) [reply test/mocking_utils.hpp:1]: If this logic gets moved to `rcpputils`, wouldn't that add an `rcpputils` test dependency to rcutils for the unit tests?

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:1]: > If this logic gets moved to rcpputils, wouldn't that add an rcpputils test dependency to rcutils for the unit tests?

It would.

> Yes, that would be a concern because if mimick_vendor becomes a runtime dependency of rcpputils/rcutils we'll have to target QL1 for mimick.

Hmm, that's a good point. And it's not in our best interest.

[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp:1]: With https://github.com/ros2/rcpputils/pull/92 implementing much of the base functionality, perhaps we can keep this more complex mocks within `rcutils`.

**Comment by hidmic** [comment]
@Blast545 @brawner PTAL at https://github.com/ros2/rcutils/pull/272/commits/9e004d7e1e11edabe0003fd77a0c81ff51d3462a, as it completely revamps how mocks are made using generic C++ code.

File: test/mocking_utils.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils.hpp]: As mentioned in https://github.com/ros2/rcpputils/pull/92, we don't want to export `mimick_vendor` as a dependency there. So we're keeping this C++ API here.

File: test/test_time.cpp [hunk 11]
```diff
@@ -188,6 +195,59 @@ TEST_F(TestTimeFixture, test_rcutils_steady_time_now) {
...
+    EXPECT_EQ(RCUTILS_RET_ERROR, ret);
+    rcutils_reset_error();
+
+    ret = rcutils_steady_time_now(&now);
+    EXPECT_EQ(RCUTILS_RET_ERROR, ret);
+    rcutils_reset_error();
+  }
+
+  {
+    auto mock = mocking_utils::patch(
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/test_time.cpp:227]: It should be, but note that these two happen in separate block statements.

File: test/test_filesystem.cpp [hunk 12]
```diff
@@ -330,22 +360,25 @@ TEST_F(TestFilesystemFixture, calculate_directory_size) {
...
+
   {
-    g_allocator.deallocate(non_existing_path, g_allocator.state);
-  });
+    auto fs = mocking_utils::patch_filesystem("lib:rcutils");
+    const char * path = "some_fake_directory/some_fake_folder";
+    fs.file_info(path).st_mode |= mocking_utils::filesystem::FileTypes::DIRECTORY;
+    fs.exhaust_file_descriptors();
+    size = rcutils_calculate_directory_size(path, g_allocator);
+    EXPECT_EQ(0u, size);
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/test_filesystem.cpp:380]: It could, but I'd keep changes to `rcutils` implementation out of this PR. [Same as before](https://github.com/ros2/rcutils/pull/272#discussion_r463206164).

File: test/test_time.cpp [hunk 13]
```diff
@@ -292,4 +372,23 @@ TEST_F(TestTimeFixture, test_rcutils_time_point_value_as_seconds_string) {
   ret = rcutils_time_point_value_as_seconds_string(&timepoint, buffer, sizeof(buffer));
   EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
   EXPECT_STREQ("-0000000000.000000100", buffer);
+
+#ifdef _WIN32
+#define vsnprintf _vsnprintf_s
+#endif
+  auto mock = mocking_utils::patch(
+    "lib:rcutils", vsnprintf, [](char * buffer, auto...) {
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/test_time.cpp:380]: I thought about it, but then not all are equal, plus [the argument](https://github.com/ros2/rcutils/pull/272#discussion_r463212224) that keeping mock behavior close to the test is better resonates with me as long as we don't have to repeat too many implementation details, like it is the case for `mocking_utils::filesystem::FileSystem`.  WDYT?

File: test/test_filesystem.cpp [hunk 14]
```diff
@@ -254,6 +277,13 @@ TEST_F(TestFilesystemFixture, is_readable_and_writable) {
     ASSERT_FALSE(nullptr == path);
     EXPECT_TRUE(rcutils_is_readable_and_writable(path));
   }
+  {
+    auto fs = mocking_utils::patch_filesystem("lib:rcutils");
+    const char * path = "fake_writable_but_non_readable_file.txt";
+    fs.file_info(path).st_mode |= mocking_utils::filesystem::Permissions::USER_READABLE;
+    fs.file_info(path).st_mode &= ~mocking_utils::filesystem::Permissions::USER_WRITABLE;
+    EXPECT_FALSE(rcutils_is_readable_and_writable(path));
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/test_filesystem.cpp:285]: Sounds good. See https://github.com/ros2/rcutils/pull/272/commits/164a36dc0860035fcb30e5e63211e273f9f75a4d.

File: test/test_time.cpp [hunk 15]
```diff
@@ -23,6 +23,13 @@
 #include "rcutils/error_handling.h"
 #include "rcutils/time.h"
 
+#include "./mocking_utils/patch.hpp"
+
+// For mocking purposes
+#ifdef _WIN32
+#define vsnprintf _vsnprintf_s
+#endif
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/test_time.cpp]: Indeed. Removed in https://github.com/ros2/rcutils/pull/272/commits/164a36dc0860035fcb30e5e63211e273f9f75a4d.

**Comment by hidmic** [comment]
CI up to `rcutils`:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11776/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6853/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9614/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11719/)

File: test/test_time.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Blast545) [reply test/test_time.cpp:222]: Oh you are right about the blocks, disregard my comment

File: test/mocking_utils/patch.hpp [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+auto make_patch(const std::string & target, std::function<SignatureT> proxy)
+{
+  return Patch<N, SignatureT>(target, proxy);
+}
+
+/// Define a dummy operator `op` for a given `type`.
+///
+/// Useful to enable patching functions that take arguments whose types
+/// do not define basic comparison operators required by Mimick.
+#define MOCKING_UTILS_DEFINE_DUMMY_OPERATOR(type, op) \
```
Comment (brawner) [inline test/mocking_utils/patch.hpp:292]: suggest: MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE

Just so when people see it in code, it's a bit clearer what this macro is doing.

File: test/test_time.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Blast545) [reply test/test_time.cpp:375]: Sounds good to me, we might consider a different approach for `rcl` if there are many repeated functions, but that's another story.

File: test/test_filesystem.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (Blast545) [reply test/test_filesystem.cpp:382]: Sounds good, I can help you opening an enhancement issue targeting that if you'd like.

File: test/test_split.cpp [hunk 17]
```diff
@@ -19,6 +19,7 @@
 #include "rcutils/error_handling.h"
 #include "rcutils/split.h"
 #include "rcutils/types/string_array.h"
+#include "rcutils/types/char_array.h"
```
Comment (brawner) [inline test/test_split.cpp:22]: Fix ordering

File: test/mocking_utils/patch.hpp [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+template<size_t N, typename SignatureT>
+struct PatchTraits;
+
+template<size_t N, typename ReturnType>
+struct PatchTraits<N, ReturnType(void)>
+{
+  mmk_mock_define(mock_type, ReturnType);
+};
+
+template<size_t N, typename ReturnType, typename ArgType0>
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils/patch.hpp]: It's not because `mmk_mock_define` is a macro that needs all the parameter types given as separate arguments at the preprocessing stage -- it won't play along with parameter packs.

File: test/mocking_utils/filesystem.hpp [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+  static constexpr mode_t USER_READABLE = S_IRUSR;
+  static constexpr mode_t USER_WRITABLE = S_IWUSR;
+#else
+  static constexpr mode_t USER_READABLE = _S_IREAD;
+  static constexpr mode_t USER_WRITABLE = _S_IWRITE;
+#endif
+};
+
+template<size_t N>
+class FileSystem
```
[the remark this answers is not in the document]
Comment (hidmic) [reply test/mocking_utils/filesystem.hpp:68]: See https://github.com/ros2/rcutils/pull/272/commits/83140ecc180b40aa9aff5e997d33cb19679d1b10.

File: test/mocking_utils/patch.hpp (hunk 16 above)
Comment (hidmic) [reply test/mocking_utils/patch.hpp]: See https://github.com/ros2/rcutils/pull/272/commits/83140ecc180b40aa9aff5e997d33cb19679d1b10.

File: test/test_split.cpp (hunk 17 above)
Comment (hidmic) [reply test/test_split.cpp]: That include actually shouldn't be there! It's a leftover, removed in https://github.com/ros2/rcutils/pull/272/commits/46a4c15b806cfe6ec1990260faed106e37763339.

File: test/mocking_utils/filesystem.hpp [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+public:
+  /// Construct mocked filesystem.
+  /**
+   * \param[in] scope Scope target string, using Mimick syntax.
+   *   \see mocking_utils::Patch documentation for further reference.
+   */
+  explicit FileSystem(const std::string & scope)
+  : opendir_mock_(MOCKING_UTILS_PATCH_TARGET(scope, opendir),
+      MOCKING_UTILS_PATCH_PROXY(opendir)),
+#ifndef _GNU_SOURCE
```
Comment (brawner) [inline test/mocking_utils/filesystem.hpp:79]: Also seems like a good candidate for combining with the ifndef block at line 88 below. Line 87 is the only line shared between the two cases.

File: test/test_filesystem.cpp [hunk 21]
```diff
@@ -201,6 +203,21 @@ TEST_F(TestFilesystemFixture, is_readable) {
     ASSERT_FALSE(nullptr == path);
     EXPECT_FALSE(rcutils_is_readable(path));
   }
+  {
+    char * path = rcutils_join_path(this->test_path, "dummy_nonexisting_file.txt", g_allocator);
```
Comment (brawner) [inline test/test_filesystem.cpp:207]:
```suggestion
    char * path = rcutils_join_path(this->test_path, "dummy_nonexistent_file.txt", g_allocator);
```

File: test/test_filesystem.cpp [hunk 22]
```diff
@@ -201,6 +203,21 @@ TEST_F(TestFilesystemFixture, is_readable) {
...
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      g_allocator.deallocate(path, g_allocator.state);
+    });
+    ASSERT_FALSE(nullptr == path);
+    EXPECT_FALSE(rcutils_is_readable(path));
+  }
+  {
+    auto fs = mocking_utils::patch_filesystem("lib:rcutils");
+    const char * path = "dummy_non_readable_file.txt";
```
Comment (brawner) [inline test/test_filesystem.cpp:217]:
```suggestion
    const char * path = "dummy_unreadable_file.txt";
```

File: test/test_filesystem.cpp [hunk 23]
```diff
@@ -232,6 +249,12 @@ TEST_F(TestFilesystemFixture, is_writable) {
     ASSERT_FALSE(nullptr == path);
     EXPECT_FALSE(rcutils_is_writable(path));
   }
+  {
+    auto fs = mocking_utils::patch_filesystem("lib:rcutils");
+    const char * path = "dummy_non_writable_file.txt";
```
Comment (brawner) [inline test/test_filesystem.cpp:254]:
```suggestion
    const char * path = "dummy_unwritable_file.txt";
```

**Comment by hidmic** [comment]
Well, it appears that there's something very peculiar about how `va_list` is implemented in `aarch64`. I don't know what the compiler is doing behind the hood, but the type does not define comparison operators nor it allows overloading such operators. As a result, functions taking `va_list` arguments are off the table in this platform :slightly_frowning_face:

I'll exclude the tests that mock `vsnprintf` in this platform.

**Comment by hidmic** [comment]
Oh well, `vsnprintf` mocking only works on Linux x86-64: binary APIs on MacOS and Windows are inaccessible and undocumented respectively, and the `va_list` type on Linux aarch64 won't play along with Mimick.

**Comment by hidmic** [comment]
Alright, finally all green. @brawner @Blast545 PTAL!

**Comment by hidmic** [comment]
Alright, thanks for the reviews guys!

**Comment by ivanpauno** [comment]
I'm not completely sure if this PR is the root cause, but there are a bunch of new test failures on the nightlies https://ci.ros2.org/view/nightly/job/nightly_linux_release/1633/.

**Comment by hidmic** [comment]
Indeed. Looks like another corner case when mocking `vsnprintf`... Looking.

**Comment by dirk-thomas** [comment]
> I'm not completely sure if this PR is the root cause, but there are a bunch of new test failures on the nightlies

Since most of them started 3 days ago (see https://ci.ros2.org/view/nightly/job/nightly_linux_thread_sanitizer/504/testReport/) I would think so.

**Comment by hidmic** [comment]
Fix PR'd in https://github.com/ros2/rcutils/pull/274.
